### PR TITLE
#161374998 Refactor Travel Calendar

### PR DESCRIPTION
**Pull request by asamisellah**
#### What does this PR do?
Improves the travel calendar filter by duration functionality and refactors travel calendar

#### Description of Task to be completed?
1. Make Travel Calendar view a component
2. Make flight details titles to reflect as expected
3. Change calendar details object-keys to camel case
4. Add react-date-range calendar
5. Create separate redux actions and saga for the download button 
6. Add success and failure toasts for the download button
7. Refactor tests

##### Prerequisites 
You should have your back-end server running
After Logging in, change your user role to  travel administrator in the database

- Using your terminal, clone this repo -``` git clone https://github.com/andela/travel_tool_front.git``` 
- Change directory to the repo in your text editor-```cd travela_tool_front```
- Open the project in your text editor
- Run ```git fetch``` on your terminal
- Checkout to this branch-```git checkout ch-refactor-travel-calendar-161374998```
- Install the dependencies-```yarn install```
- Start the server-```yarn start ```
- Navigate to localhost:3000 in your browser
- Login to the application
- Navigate to the dashboard on the side nav
- Scroll down the page to the Travel Calendar
- Click on the download button to download a CSV file of the travel calendar data
- Click on the drop-down button to filter travel calendar details by duration

#### Any background context you want to provide?
Currently, the travel calendar analytics is being fetched from a mocked API endpoint

#### What are the relevant pivotal tracker stories?
[#161374998](https://www.pivotaltracker.com/n/projects/2184887/stories/161374998)

#### Screenshots (if appropriate)
screen shot 2018-11-13 at 11 53 16

screen shot 2018-11-13 at 11 53 30

screen shot 2018-11-13 at 11 54 00

### CODE REVIEW

**Comment by asamisellah** [comment]
Hi guys, I am aware of the test coverage... I'm working on that. Kindly review the functionality for now. Thanks in advance

File: src/components/TravelCalendar/index.jsx [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        filterBtnLabel: label
+      }));
+      fetchCalendarAnalytics({type:'json', filter:query});
+    }
+  }
+
+  handleFilterBtn = () => {
+    const {isCalendarOpen} = this.state;
+    isCalendarOpen ? this.setState(prevState=>({...prevState, isCalendarOpen: false}))
+      : this.setState(prevState=>({...prevState, isCalendarOpen: true}));
```
[the remark this answers is not in the document]
Comment (asamisellah) [reply src/components/TravelCalendar/index.jsx]: Right, that makes sense.

File: src/components/TravelCalendar/__tests__/TravelCalendar.test.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    btn.simulate('click');
+    expect(props.downloadCalendarAnalytics).toHaveBeenCalled()
+  });
+
+  it('checks that spinner is rendered when isLoading is true', () => {
+    wrapper.setProps(travelCalendar.isLoading = true);
+    const spinner = wrapper.find('.analyticsReport__report-details')
+    expect(spinner.length).toBe(1)
+  });
+})
```
Comment (ronaldndirangu) [inline src/components/TravelCalendar/__tests__/TravelCalendar.test.js]: Please add an end of line here.

File: src/redux/reducers/__tests__/travelCalendar.test.js [hunk 3]
```diff
@@ -48,4 +53,15 @@ describe('Travel Calendar Reducer', ()=>{
...
+    const action = {
+      type: DOWNLOAD_CALENDAR_ANALYTICS_FAILURE,
+      error: 'Download Unsuccessful'
+    };
+    expect(travelCalendarReducer(initialState, action)).toEqual({
+      ...initialState,
+      downloadError: 'Download Unsuccessful'
+    });
+  });
 });
```
Comment (ronaldndirangu) [inline src/redux/reducers/__tests__/travelCalendar.test.js]: Add an end of line here also

File: package.json [hunk 4]
```diff
@@ -69,6 +70,7 @@
     "prop-types": "^15.6.2",
     "raf": "3.4.0",
     "react": "^16.4.1",
+    "react-date-range": "^1.0.0-beta",
     "react-datepicker": "^1.6.0",
```
Comment (ronaldndirangu) [inline package.json]: Do we need the react-datepicker here?

**Comment by asamisellah** [comment]
@davenjeru I'm currently still mocking the backend, therefore the file type being returned if a json file and not a csv file. This will be fixed once the backend is complete.

File: package.json (hunk 4 above)
Comment (asamisellah) [reply package.json]: Yes, there are other components like the request form that are using the package

Comment (ronaldndirangu) [reply package.json]: Okay.

**Comment by Stylll** [comment]
LGTM

File: src/components/TravelCalendar/__tests__/TravelCalendar.test.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
+import React from 'react';
+import { shallow, mount } from 'enzyme';
+import sinon from 'sinon';
```
Comment (amarachukwu-agbo) [inline src/components/TravelCalendar/__tests__/TravelCalendar.test.js]: Please remove unused modules imported on line 3 - 6 and 'mount' too

File: src/redux/middleware/travelCalendarSaga.js [hunk 6]
```diff
@@ -23,5 +24,21 @@ export function* fetchCalendarAnalyticsSaga(action) {
...
 
 export function* watchFetchCalendarAnalytics() {
-  yield takeEvery(FETCH_CALENDAR_ANALYTICS, fetchCalendarAnalyticsSaga);
+  yield takeLatest(FETCH_CALENDAR_ANALYTICS, fetchCalendarAnalyticsSaga);
+}
+
+export function* downloadCalendarAnalyticsSaga(action) {
+  try {
+    const response = yield call(CalendarAnalyticsAPI.getCalendarAnalytics, action.query);
+    yield FileSaver.saveAs(response.data, 'Travel Calendar Analytics');
```
Comment (amarachukwu-agbo) [inline src/redux/middleware/travelCalendarSaga.js]: Please add toast for successful download

**Comment by amarachukwu-agbo** [comment]
Good job @asamisellah. I noticed that the travel calendar does not close when you click outside. @DennisWanjiru's implementation closes the calendar when there is a click outside. I think this is more intuitive as the user will not struggle to figure out how to close the calendar.
Since these two calendars are on the same page, I think it is best to maintain consistency and stick with Dennis' implementation.
Also, check that the right font-size and color is being used in the calendar input that holds the selected date range

**Comment by samuelkubai** [comment]
@asamisellah what's the progress on this?

**Comment by asamisellah** [comment]
@samuelkubai I'm done implementing the feedback from Amara. It's ready for review

**Pull request merged**